frontends/p4/typeChecking/typeChecker.cpp

```diff
@@ -1755,6 +1755,37 @@ const IR::Node* TypeInference::preorder(IR::EntriesList* el) {
     return el;
 }
 
+// If e->type is Type_Bits, this method doesn't do anything.
+const IR::Node* TypeInference::preorder(IR::Type_SerEnum* se) {
```

Comment:
This code should be moved into the postorder for SerEnum.
Then it becomes much simpler; you do `typeMap->getTypeType(se->type)` and check that it is `bit<>` or `int<>` and give an error otherwise. Also, your error messages are not clear, the error should be "type %1% cannot be used for the underlying type of an enum"

Reply:
It's tricky to change the postorder method because without preorder, the SerEnum input into the postorder method has type as Type_Name.  We would have to morph the SerEnum to be of Type_Bits.  If you see the preorder method, a new SerEnum is allocated to change the type.  

I have sent detailed logs in an email.

Reply:
This was the note in an email to @mbudiu-vmw.

Its tricky to remove the preorder and add code to postorder SerEnum.  These are the diffs for postorder as per your comment.

```p4
diff --git a/frontends/p4/typeChecking/typeChecker.cpp b/frontends/p4/typeChecking/typeChecker.cpp
index 3256483a..c0070473 100644
--- a/frontends/p4/typeChecking/typeChecker.cpp
+++ b/frontends/p4/typeChecking/typeChecker.cpp
@@ -1223,7 +1223,14 @@ const IR::Node* TypeInference::postorder(IR::Type_Enum* type) {
     return type;
}

const IR::Node* TypeInference::postorder(IR::Type_SerEnum* type) {
+    auto t = typeMap->getTypeType(type->type, true);
+    CHECK_NULL(t);
+    if (!t->is<IR::Type_Bits>()) {
+        typeError("%1%: cannot use type %2%", type, t);
+        return type;
+    }
     auto canon = setTypeType(type);
     for (auto e : *type->getDeclarations())
         setType(e->getNode(), canon);
```
But when I test with `issue2391.p4` which has a SerEnum using tydedef, The “type” coming in to the postorder method has type as Type_Name.  This is why I didn’t change rest of the postorder code using “type” because the “type” is incorrect.  

```p4
[2126] Type_SerEnum name=EthTypes declid=11
  annotations: [3] Annotations
  type: [171] Type_Name
    path: [170] Path name=EthT absolute=0
  [2132] SerEnumMember name=IPv4 declid=25
    value: [2139] Constant value=2048 base=16
      type: [151] Type_Bits size=16 isSigned=0
  [2142] SerEnumMember name=ARP declid=26
    value: [2149] Constant value=2054 base=16
      type: [151] Type_Bits size=16 isSigned=0...
  [2152] SerEnumMember name=RARP declid=27
    value: [2159] Constant value=32821 base=16
      type: [151] Type_Bits size=16 isSigned=0...
  [2162] SerEnumMember name=EtherTalk declid=28
    value: [2169] Constant value=32923 base=16
      type: [151] Type_Bits size=16 isSigned=0...
  [2172] SerEnumMember name=VLAN declid=29
    value: [2179] Constant value=33024 base=16
      type: [151] Type_Bits size=16 isSigned=0...
  [2182] SerEnumMember name=IPX declid=30
    value: [2189] Constant value=33079 base=16
      type: [151] Type_Bits size=16 isSigned=0...
  [2192] SerEnumMember name=IPv6 declid=31
    value: [2199] Constant value=34525 base=16
      type: [151] Type_Bits size=16 isSigned=0...
```

This is why I was morphing the SerEnum with type of Type_Name in preorder with a “new” allocation of SerEnum.  If I add a new allocation of SerEnum in postorder, I get an error saying 

```bash
“In file: /home/hemant/int/p4c/frontends/p4/typeChecking/typeChecker.cpp:144
[31mCompiler Bug[0m: At this point in the compilation typechecking should not infer new types anymore, but it did.”
```

In summary, I think we could keep the preorder Serenum method.